### PR TITLE
rails4 master merge

### PR DESCRIPTION
**Pull request by komidore64**

### CODE REVIEW

**Comment by johnpmitsch** [comment]
@komidore64 It looks fine to me, @cpeters can you double check my review and ack?

File: app/models/katello/glue/candlepin/pool.rb [hunk 1]
```diff
@@ -6,179 +6,124 @@ def self.included(base)
...
+      def create_activation_key_associations
         keys = Resources::Candlepin::ActivationKey.get(nil, "?include=id&include=pools.pool.id")
         activation_key_ids = keys.collect do |key|
           key['id'] if key['pools'].any? { |pool| pool['pool']['id'] == cp_id }
         end
-
-        return Katello::ActivationKey.where(:cp_id => activation_key_ids.compact)
+        related_keys = ::Katello::ActivationKey.where(:cp_id => activation_key_ids.compact)
+        related_keys.each do |key|
+          Katello::PoolActivationKey.find_or_create_by_activation_key_id_and_pool_id(key.id, self.id)
```
Comment (cpeters) [inline app/models/katello/glue/candlepin/pool.rb]: Is this a defined method or a rails3 AR method missing?

File: app/models/katello/glue/candlepin/subscription.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      end
+
+      def create_product_associations
+        products = self.query_products
+        cp_product_ids = products.map { |product| product["id"] }
+        cp_product_ids << self.product_id if self.product_id
+        cp_product_ids.each do |cp_id|
+          product = ::Katello::Product.where(:cp_id => cp_id)
+          if product.any?
+            ::Katello::SubscriptionProduct.find_or_create_by_subscription_id_and_product_id(self.id, product.first.id)
```
Comment (cpeters) [inline app/models/katello/glue/candlepin/subscription.rb]: Another method that might be deprecated

File: app/models/katello/glue/candlepin/pool.rb (hunk 1 above)
Comment (johnpmitsch) [inline app/models/katello/glue/candlepin/pool.rb]: this is an AR 3 method and will have to be updated

File: app/controllers/katello/api/v2/subscriptions_controller.rb [hunk 3]
```diff
@@ -20,22 +17,32 @@ class Api::V2::SubscriptionsController < Api::V2::ApiController
...
 
-      respond_for_index(:collection => subscriptions)
+    def index_relation
+      return available_for_system if params[:available_for] == "content_host"
+      return available_for_activation_key if params[:available_for] == "activation_key"
+      collection = Pool.readable
+      collection = collection.where(:id => ActivationKey.find(params[:activation_key_id]).pools) if params[:activation_key_id]
+      collection = collection.get_for_organization(Organization.find(params[:organization_id])) if params[:organization_id]
+      if params[:system_id]
+        pool_ids = System.find_by_uuid(params[:system_id]).pools.map { |x| x['id'] }
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/subscriptions_controller.rb]: I'll comment what I see needs to be updated like here

File: app/controllers/katello/api/v2/subscriptions_controller.rb [hunk 4]
```diff
@@ -174,6 +185,26 @@ def manifest_history
...
+    param :match_installed, :bool, :desc => N_("Return subscriptions that match installed products")
+    param :no_overlap, :bool, :desc => N_("Return subscriptions which do not overlap with a currently-attached subscription")
+    def available_for_system
+      params[:match_system] = ::Foreman::Cast.to_bool(params[:match_system]) if params[:match_system]
+      params[:match_installed] = ::Foreman::Cast.to_bool(params[:match_installed]) if params[:match_installed]
+      params[:no_overlap] = ::Foreman::Cast.to_bool(params[:no_overlap]) if params[:no_overlap]
+      pools = @system.filtered_pools(params[:match_system], params[:match_installed],
+                                     params[:no_overlap])
+      if pools
+        available = pools.collect { |cp_pool| ::Katello::Pool.find_by_cp_id(cp_pool['id']) }
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/subscriptions_controller.rb]: update

File: app/lib/actions/candlepin/import_pool_handler.rb [hunk 5]
```diff
@@ -44,24 +44,25 @@ def handle(message)
 
       private
 
-      def reindex_or_unindex_pool(pool_id)
-        ::Katello::Pool.find_pool(pool_id)
-        reindex_pool(pool_id)
+      def import_or_remove_pool(pool_id)
+        ::Katello::Pool.find_by_cp_id(pool_id).import_data
```
Comment (johnpmitsch) [inline app/lib/actions/candlepin/import_pool_handler.rb]: update

File: app/lib/actions/candlepin/import_pool_handler.rb [hunk 6]
```diff
@@ -44,24 +44,25 @@ def handle(message)
...
+        ::Katello::Pool.import_pool(pool_id)
       rescue RestClient::ResourceNotFound
         @logger.debug "skipped re-index of non-existent pool #{pool_id}"
       end
 
-      def unindex_pool(pool_id)
-        @logger.info "removing pool from index #{pool_id}."
-        ::Katello::Pool.remove_from_index(pool_id)
+      def remove_pool(pool_id)
+        pool = ::Katello::Pool.find_by_cp_id(pool_id)
```
Comment (johnpmitsch) [inline app/lib/actions/candlepin/import_pool_handler.rb]: update

File: app/models/katello/activation_key.rb [hunk 7]
```diff
@@ -104,20 +95,12 @@ def products
...
       if cp_pools
-        pools = cp_pools.collect { |cp_pool| Pool.find_pool(cp_pool['id'], cp_pool) }
-        product_ids = pools.map(&:product_id)
-        marketing_products = MarketingProduct.includes(:engineering_products, :marketing_engineering_products).
-            where(:cp_id => product_ids)
-        products = Product.where(:cp_id => product_ids).where('type != ?', "Katello::MarketingProduct")
-
-        marketing_products.each do |product|
-          all_products += product.engineering_products
+        pools = cp_pools.collect { |cp_pool| Pool.find_by_cp_id(cp_pool['id']) }
```
Comment (johnpmitsch) [inline app/models/katello/activation_key.rb]: update

File: app/models/katello/glue/candlepin/consumer.rb [hunk 8]
```diff
@@ -420,18 +410,20 @@ def products
         all_products = []
 
         self.entitlements.each do |entitlement|
-          pool = Katello::Pool.find_pool(entitlement['pool']['id'])
+          pool = Katello::Pool.find_by_cp_id(entitlement['pool']['id'])
```
Comment (johnpmitsch) [inline app/models/katello/glue/candlepin/consumer.rb]: update

File: app/models/katello/glue/candlepin/subscription.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        subscription_attributes = {}
+        subscription_json = self.backend_data
+
+        subscription_json["product"]["attributes"].map { |attr| subscription_attributes[attr["name"].to_sym] = attr["value"] }
+
+        subscription_attributes[:name] = subscription_json["product"]["name"]
+        subscription_attributes[:product_id] = subscription_json["product"]["id"]
+        subscription_attributes[:instance_multiplier] = subscription_json["product"]["multiplier"]
+        subscription_attributes[:stacking_id] = subscription_json["stackId"]
+        organization = Organization.find_by_label(subscription_json["owner"]["key"]) if subscription_json["owner"]
```
Comment (johnpmitsch) [inline app/models/katello/glue/candlepin/subscription.rb]: update (unless this is a declared Organization method)

File: test/actions/import_pool_test.rb [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
+  describe ::Actions::Candlepin::ImportPoolHandler do
     let(:pool_id) { '12345' }
     let(:pool) { OpenStruct.new(:id => pool_id) }
 
     before do
-      ::Actions::Candlepin::ReindexPoolSubscriptionHandler.any_instance.stubs(:attempt_find_pool).returns('id' => pool_id)
-      ::Katello::Pool.stubs(:find_pool).returns(pool)
-      ::Katello::Pool.stubs(:index_pools)
+      ::Actions::Candlepin::ImportPoolHandler.any_instance.stubs(:attempt_find_pool).returns('id' => pool_id)
+      ::Katello::Pool.stubs(:find_by_cp_id).returns(pool)
```
Comment (johnpmitsch) [inline test/actions/import_pool_test.rb]: update

File: test/models/pool_test.rb [hunk 11]
```diff
@@ -36,19 +41,107 @@ def test_recently_expired_does_not_get_long_expired_subscriptions
...
-    end
-
     def test_systems
       active_pool = FactoryGirl.build(:katello_pool, :active)
       systems = [katello_systems(:simple_server)]
       System.expects(:all_by_pool).with(active_pool.cp_id).returns(systems)
       assert_equal active_pool.systems, systems
     end
+
+    def test_return_find_by_cp_id
```
Comment (johnpmitsch) [inline test/models/pool_test.rb]: update

File: app/models/katello/glue/candlepin/candlepin_object.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      end
+
+      def get_candlepin_ids(organization)
+        self.get_for_owner(organization).map { |subscription| subscription["id"] }
+      end
+
+      def import_candlepin_ids(organization)
+        candlepin_ids = self.get_candlepin_ids(organization)
+        candlepin_ids.each do |cp_id|
+          self.find_or_create_by_cp_id(cp_id) unless cp_id.nil?
```
Comment (cpeters) [inline app/models/katello/glue/candlepin/candlepin_object.rb]: possibly another rails3 specific method; however I do remember seeing stuff related to candlepin where theses types of methods were overridden.

File: app/models/katello/glue/candlepin/pool.rb [hunk 13]
```diff
@@ -6,179 +6,124 @@ def self.included(base)
...
-      def find_by_id(pool_id)
-        Katello::Pool.find_by(:cp_id => pool_id) || Pool.new(Resources::Candlepin::Pool.find(pool_id))
-      end
-
-      def find_by_id!(pool_id)
-        subscription = find_by_id(pool_id)
-        fail ActiveRecord::RecordNotFound if subscription.nil?
-        subscription
+      def import_pool(cp_pool_id)
+        pool = Katello::Pool.find_or_create_by_cp_id(cp_pool_id)
```
Comment (cpeters) [inline app/models/katello/glue/candlepin/pool.rb]: another one

File: app/models/katello/glue/candlepin/candlepin_object.rb (hunk 12 above)
Comment (johnpmitsch) [inline app/models/katello/glue/candlepin/candlepin_object.rb]: forgot about those :/ ya that is a rails 3 specific method and can be updated

File: db/seeds.rb [hunk 14]
```diff
@@ -56,8 +56,6 @@ def format_errors(model = nil)
   template.locations << Location.first unless template.locations.include?(Location.first) || Location.count.zero?
 end
 
-Katello::Util::Search.backend_search_classes.each { |c| c.create_index }
-
 # Proxy features
 feature = Feature.find_or_create_by(:name => 'Pulp')
```
Comment (johnpmitsch) [inline db/seeds.rb]: another

File: app/models/katello/glue/candlepin/pool.rb (hunk 1 above)
Comment (komidore64) [inline app/models/katello/glue/candlepin/pool.rb]: incoming

File: app/models/katello/glue/candlepin/subscription.rb (hunk 2 above)
Comment (komidore64) [inline app/models/katello/glue/candlepin/subscription.rb]: incoming

File: app/controllers/katello/api/v2/subscriptions_controller.rb (hunk 3 above)
Comment (komidore64) [inline app/controllers/katello/api/v2/subscriptions_controller.rb]: incoming

File: app/controllers/katello/api/v2/subscriptions_controller.rb (hunk 4 above)
Comment (komidore64) [inline app/controllers/katello/api/v2/subscriptions_controller.rb]: incoming

File: app/lib/actions/candlepin/import_pool_handler.rb (hunk 5 above)
Comment (komidore64) [inline app/lib/actions/candlepin/import_pool_handler.rb]: incoming

File: app/lib/actions/candlepin/import_pool_handler.rb (hunk 6 above)
Comment (komidore64) [inline app/lib/actions/candlepin/import_pool_handler.rb]: incoming

File: app/models/katello/activation_key.rb (hunk 7 above)
Comment (komidore64) [inline app/models/katello/activation_key.rb]: incoming

File: app/models/katello/glue/candlepin/consumer.rb (hunk 8 above)
Comment (komidore64) [inline app/models/katello/glue/candlepin/consumer.rb]: incoming

File: app/models/katello/glue/candlepin/subscription.rb (hunk 9 above)
Comment (komidore64) [inline app/models/katello/glue/candlepin/subscription.rb]: incoming

File: test/actions/import_pool_test.rb (hunk 10 above)
Comment (komidore64) [inline test/actions/import_pool_test.rb]: incoming

File: test/models/pool_test.rb (hunk 11 above)
Comment (komidore64) [inline test/models/pool_test.rb]: incoming

File: app/models/katello/glue/candlepin/candlepin_object.rb (hunk 12 above)
Comment (komidore64) [inline app/models/katello/glue/candlepin/candlepin_object.rb]: incoming

File: app/models/katello/glue/candlepin/pool.rb (hunk 13 above)
Comment (komidore64) [inline app/models/katello/glue/candlepin/pool.rb]: incoming

File: db/seeds.rb (hunk 14 above)
Comment (komidore64) [inline db/seeds.rb]: isn't this one correct?

Comment (cpeters) [inline db/seeds.rb]: That should work afaik

Comment (johnpmitsch) [inline db/seeds.rb]: ya it is heh

**Comment by johnpmitsch** [comment]
@komidore64 ACK

**Comment by cpeters** [comment]
ACK

- Chris

**Pull request merged**